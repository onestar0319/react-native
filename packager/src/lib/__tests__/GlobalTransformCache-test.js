/**
 * Copyright (c) 2016-present, Facebook, Inc.
 * All rights reserved.
 *
 * This source code is licensed under the BSD-style license found in the
 * LICENSE file in the root directory of this source tree. An additional grant
 * of patent rights can be found in the PATENTS file in the same directory.
 */

'use strict';

jest.disableAutomock();
jest.useRealTimers();

const mockFetch = jest.fn();
jest.mock('node-fetch', () => mockFetch);

const {URIBasedGlobalTransformCache} = require('../GlobalTransformCache');
const FetchError = require('node-fetch/lib/fetch-error');
const path = require('path');

async function fetchResultURIs(keys: Array<string>): Promise<Map<string, string>> {
  return new Map(keys.map(key => [key, `http://globalcache.com/${key}`]));
}

async function fetchResultFromURI(uri: string): Promise<?CachedResult> {
  return {
    code: `/* code from ${uri} */`,
    dependencies: [],
    dependencyOffsets: [],
  };
}

describe('GlobalTransformCache', () => {

  it('fetches results', async () => {
    const cache = new URIBasedGlobalTransformCache({
      fetchResultFromURI,
      fetchResultURIs,
      profiles: [{dev: true, minify: false, platform: 'ios'}],
      rootPath: __dirname,
      storeResults: null,
    });
    const transformOptions = {
      dev: true,
      minify: false,
      platform: 'ios',
      transform: {
        generateSourceMaps: false,
        dev: false,
        hot: false,
        inlineRequires: false,
        platform: 'ios',
        projectRoot: path.join(__dirname, 'root'),
      },
    };
    const result = await Promise.all([cache.fetch({
      filePath: 'foo.js',
      sourceCode: '/* beep */',
      getTransformCacheKey: () => 'abcd',
      transformOptions,
    }), cache.fetch({
      filePath: 'bar.js',
      sourceCode: '/* boop */',
      getTransformCacheKey: () => 'abcd',
      transformOptions,
    })]);
    expect(result).toMatchSnapshot();
  });

  describe('fetchResultFromURI', () => {

    const defaultFetchMockImpl = async uri => ({
      status: 200,
      json: async () => ({
        code: `/* code from ${uri} */`,
        dependencies: [],
        dependencyOffsets: [],
      }),
    });

    beforeEach(() => {
      mockFetch.mockReset();
    });

    it('fetches result', async () => {
      mockFetch.mockImplementation(defaultFetchMockImpl);
      const result = await URIBasedGlobalTransformCache
        .fetchResultFromURI('http://globalcache.com/foo');
      expect(result).toMatchSnapshot();
    });

    it('retries once on timeout', async () => {
      mockFetch.mockImplementation(async uri => {
        mockFetch.mockImplementation(defaultFetchMockImpl);
        throw new FetchError('timeout!', 'request-timeout');
      });
      const result = await URIBasedGlobalTransformCache
        .fetchResultFromURI('http://globalcache.com/foo');
      expect(result).toMatchSnapshot();
    });

  });

});
