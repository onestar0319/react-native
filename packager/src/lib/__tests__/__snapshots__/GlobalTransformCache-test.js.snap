// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GlobalTransformCache fetchResultFromURI fetches result 1`] = `
Object {
  "code": "/* code from http://globalcache.com/foo */",
  "dependencies": Array [],
  "dependencyOffsets": Array [],
}
`;

exports[`GlobalTransformCache fetchResultFromURI retries once on timeout 1`] = `
Object {
  "code": "/* code from http://globalcache.com/foo */",
  "dependencies": Array [],
  "dependencyOffsets": Array [],
}
`;

exports[`GlobalTransformCache fetches results 1`] = `
Array [
  Object {
    "code": "/* code from http://globalcache.com/cd6df9b7e86839dafc5e9b5b493a28cbc55074e7-foo.js */",
    "dependencies": Array [],
    "dependencyOffsets": Array [],
  },
  Object {
    "code": "/* code from http://globalcache.com/6329a317fcf94d74cc9a1de6442ee7c25e27a507-bar.js */",
    "dependencies": Array [],
    "dependencyOffsets": Array [],
  },
]
`;
